### PR TITLE
Enhance Tubes

### LINKED ISSUE #59: Sock#puts without argument should print an empty line

### LINKED ISSUE #60: Sock#interact should catch EOFError

### LINKED ISSUE #61: Sock#puts should not append newline if already presented

### PR DESCRIPTION
**Pull request by hanhanW**
* Implement gets for Tubes
* Implement puts for Tubes
  * Fixed #59 
  * Fixed #61 
* Catch EOFError in the interative mode ( Fixed #60 )
* The sending methods in Tubes return the number of bytes had been sent.

---
This change is [Reviewable](https://reviewable.io/reviews/peter50216/pwntools-ruby/63)

### CODE REVIEW

File: lib/pwnlib/tubes/tube.rb [hunk 1]
```diff
@@ -198,27 +221,49 @@ def recvregex(regex, timeout: nil)
...
+      #   The objects to send.
+      #
+      # @!macro send_return_definition
+      def puts(*objs)
+        return write(context.newline) if objs.empty?
+        objs = *objs.flatten
+        s = ''
+        objs.map(&:to_s).each do |elem|
+          s << elem
+          s << context.newline if elem.empty? || !elem.end_with?(context.newline)
```
Comment (ShikChen) [inline lib/pwnlib/tubes/tube.rb]: Why need to check `elem.empty?`?

File: lib/pwnlib/tubes/tube.rb [hunk 2]
```diff
@@ -236,6 +281,8 @@ def interact
             $stdout.write(s)
           end
         end
+      rescue
```
Comment (ShikChen) [inline lib/pwnlib/tubes/tube.rb]: Should only rescue `EOFError`.

File: lib/pwnlib/tubes/tube.rb [hunk 3]
```diff
@@ -9,10 +9,18 @@
 module Pwnlib
   module Tubes
     # Things common to all tubes (sockets, tty, ...)
+    # @!macro [new] drop_definition
+    #   @param [Boalean] drop
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: Boolean

File: lib/pwnlib/tubes/tube.rb [hunk 4]
```diff
@@ -183,7 +193,20 @@ def recvuntil(delims, drop: false, timeout: nil)
...
+      # Receives the next "line" from the tube; lines are separated by +sep+.
+      # The difference with IO#gets is using +context.newline+ as default newline.
+      #
+      # @param [String] sep
+      #   The separator.
+      # @!macro drop_definition
+      # @!macro timeout_definition
+      #
+      # @return [String]
+      #   The next "line".
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: add examples to clarify its recvuntil-like functionality

File: lib/pwnlib/tubes/tube.rb (hunk 2 above)
Comment (david942j) [reply lib/pwnlib/tubes/tube.rb]: Future work: Use our own Exception class

**Comment by hanhanW** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[lib/pwnlib/tubes/tube.rb, line 13 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r1-13:-KxrFMuVCUvol7kdrJc0:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/23a7ccaff555bed4decb7f18fa36aeef2ab8bece/lib/pwnlib/tubes/tube.rb#L13)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

Boolean
</blockquote></details>

Done.

---

*[lib/pwnlib/tubes/tube.rb, line 206 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r1-206:-KxrWDcaTKyAnjg83qZU:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/23a7ccaff555bed4decb7f18fa36aeef2ab8bece/lib/pwnlib/tubes/tube.rb#L206)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

add examples to clarify its recvuntil-like functionality
</blockquote></details>

Done.

---

*[lib/pwnlib/tubes/tube.rb, line 263 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r1-263:-KxrFs0kc6Cr3jeRJPdR:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/23a7ccaff555bed4decb7f18fa36aeef2ab8bece/lib/pwnlib/tubes/tube.rb#L263)):*
<details><summary><i>Previously, ShikChen wrote…</i></summary><blockquote>

Why need to check `elem.empty?`?
</blockquote></details>

Done.

---

*[lib/pwnlib/tubes/tube.rb, line 284 at r1](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r1-284:-KxrHg8WyGOVZy4Sys-R:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/23a7ccaff555bed4decb7f18fa36aeef2ab8bece/lib/pwnlib/tubes/tube.rb#L284)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

Future work: Use our own Exception class 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63)*

**Comment by hanhanW** [comment]
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[lib/pwnlib/tubes/tube.rb, line 198 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r2-198:-Kxw1pNhfdOEzGwhv0C5:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/02e077b408107180cc2cb304f551a7d0df3ef434/lib/pwnlib/tubes/tube.rb#L198)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

+IO#gets+
</blockquote></details>

Done.

---

*[lib/pwnlib/tubes/tube.rb, line 201 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r2-201:-Kxw4lolCiheB5CFktxI:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/02e077b408107180cc2cb304f551a7d0df3ef434/lib/pwnlib/tubes/tube.rb#L201)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

describe the behavior when an integer is passed
</blockquote></details>

Done.

---

*[lib/pwnlib/tubes/tube.rb, line 274 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r2-274:-KxwDkzluAM7splOOK-d:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/02e077b408107180cc2cb304f551a7d0df3ef434/lib/pwnlib/tubes/tube.rb#L274)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

+IO#puts+
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63)*

File: lib/pwnlib/tubes/tube.rb [hunk 5]
```diff
@@ -198,27 +249,69 @@ def recvregex(regex, timeout: nil)
...
+      # @param [Array<Object>] objs
+      #   The objects to send.
+      #
+      # @!macro send_return_definition
+      #
+      # @example
+      #   s.puts
+      #   puts client.recv
+      #   #
+      #
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: add `#=> nil`, otherwise it looks weird

File: lib/pwnlib/tubes/tube.rb [hunk 6]
```diff
@@ -198,27 +249,69 @@ def recvregex(regex, timeout: nil)
...
+      def sendline(obj)
+        s = obj.to_s + context.newline
+        write(s)
+      end
+
+      # Sends the given object(s) to the tube.
+      # The difference with +IO#puts+ is using +context.newline+ as default newline.
+      #
+      # @param [Array<Object>] objs
+      #   The objects to send.
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: to be sent

File: lib/pwnlib/tubes/tube.rb [hunk 7]
```diff
@@ -169,12 +180,11 @@ def recvuntil(delims, drop: false, timeout: nil)
         end
       end
 
-      # Receive a single line from the tube.
+      # Receives a single line from the tube.
       # A "line" is any sequence of bytes terminated by the byte sequence set in +context.newline+,
       # which defaults to +"\n"+.
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: `+"\n"+` broken in parsed doc

**Comment by hanhanW** [comment]
Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[lib/pwnlib/tubes/tube.rb, line 185 at r3](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r3-185:-KxwmAg8SP0MQ21-dYI5:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/84a7539673cf8135d518d914e0598c9b76aaf215/lib/pwnlib/tubes/tube.rb#L185)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

`+"\n"+` broken in parsed doc
</blockquote></details>

Done.

---

*[lib/pwnlib/tubes/tube.rb, line 282 at r3](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r3-282:-KxwmHbXTQ_1e-5gjaRg:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/84a7539673cf8135d518d914e0598c9b76aaf215/lib/pwnlib/tubes/tube.rb#L282)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

to be sent
</blockquote></details>

Done.

---

*[lib/pwnlib/tubes/tube.rb, line 290 at r3](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-KxqtbzVB3T-5IZpNB6--r3-290:-KxwmN-ZPeGaqBp9yjo5:b-896fix) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/84a7539673cf8135d518d914e0598c9b76aaf215/lib/pwnlib/tubes/tube.rb#L290)):*
<details><summary><i>Previously, david942j (david942j) wrote…</i></summary><blockquote>

add `#=> nil`, otherwise it looks weird
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63)*

**Comment by david942j** [comment]
@ShikChen

**Comment by david942j** [comment]
Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63)*

**Comment by david942j** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/63#-:-KyRON8dRA9c70uF3rJi:bnfp4nl)*

**Pull request merged**